aiopg/utils.py

```diff
@@ -193,37 +197,97 @@ def __aexit__(self, exc_type, exc_val, exc_tb):
 class _PoolCursorContextManager:
     """Context manager.
 
-    This enables the following idiom for acquiring and releasing a
+    This enables the following idioms for acquiring and releasing a
     cursor around a block:
 
         with (yield from pool.cursor()) as cur:
             yield from cur.execute("SELECT 1")
 
+        async with pool.cursor() as cur:
+            yield from cur.execute("SELECT 1")
+
     while failing loudly when accidentally using:
 
         with pool:
             <block>
     """
 
-    __slots__ = ('_pool', '_conn', '_cur')
+    __slots__ = ('_pool', '_cursor_kwargs', '_cur')
 
-    def __init__(self, pool, conn, cur):
+    def __init__(self, pool, cursor_kwargs=None):
         self._pool = pool
-        self._conn = conn
-        self._cur = cur
+        self._cursor_kwargs = cursor_kwargs
+        self._cur = None
 
     def __enter__(self):
         return self._cur
 
     def __exit__(self, *args):
         try:
             self._cur.close()
-            self._pool.release(self._conn)
+            self._pool.__exit__(*args)
         finally:
-            self._pool = None
-            self._conn = None
             self._cur = None
 
+    @asyncio.coroutine
+    def _init_pool(self, with_aenter):
+        assert not self._cur
+
+        if with_aenter:
+            conn = None
+        else:
+            conn = yield from self._pool.acquire()
+
+        # self._pool now morphs into a _PoolConnectionContextManager
+        self._pool = _PoolConnectionContextManager(self._pool, conn)
+
+    @asyncio.coroutine
+    def _yield_await_init(self):
+        yield from self._init_pool(False)
+        self._cur = yield from self._pool.conn.cursor(**self._cursor_kwargs)
+        return self
+
+    def __iter__(self):
+        # This will get hit if you use "yield from pool.cursor()"
+        if PY_35:
```

Comment:
this is just so they to start using `async with pool.cursor()`  instead of `with (yield from pool.cursor())` so that the cursor will be actually closed at the end of the context.